pkg/codegen/docs/gen.go

```diff
@@ -1753,61 +1757,194 @@ func (mod *modContext) getModuleFileName() string {
 	return mod.mod
 }
 
+type moduleConflictResolver struct {
+	dctx *docGenContext
+	seen map[string]struct{}
+}
+
+func (dctx *docGenContext) newModuleConflictResolver() moduleConflictResolver {
+	return moduleConflictResolver{
+		dctx: dctx,
+		seen: map[string]struct{}{},
+	}
+}
+
+func (r *moduleConflictResolver) getSafeName(name string, item interface{}) string {
+
+	candidate, alreadyResolved := r.dctx.moduleConflictLinkMap[item]
+	if alreadyResolved {
+		return candidate
+	}
+	var prefixes []string
+	switch item.(type) {
+	case *schema.Resource:
+		prefixes = []string{"", "res-"}
+	case *schema.Function:
+		prefixes = []string{"", "fn-"}
+	case *modContext:
+		prefixes = []string{"", "mod-"}
+	default:
+		prefixes = []string{""}
+	}
+	for _, prefix := range prefixes {
+		candidate := prefix + name
+		if _, exists := r.seen[candidate]; exists {
+			continue
+		}
+		r.seen[candidate] = struct{}{}
+		r.dctx.moduleConflictLinkMap[item] = candidate
+		return candidate
+	}
+
+	glog.Error("skipping unresolvable duplicate file name: ", name)
+	return ""
```

Comment:
We don't actually check getSafeName for it returning "". Should this just do the current behavior of the clashing name.

Reply:
Oh, good point. I missed this. @dixler I think each `getSafeName` should continue the loop if empty.